xml_converter/src/xml_converter.cpp

```diff
@@ -32,47 +33,121 @@
+    vector<CategoryWithinSinglePack> categories;
+};
+
+map<UniqueId, CategoryWithinSinglePack> read_taco_directory(
     string input_path,
     map<string, Category>* marker_categories,
     vector<Parseable*>* parsed_pois
 ) {
-    map<string, vector<string>> top_level_category_file_locations;
+    map<UniqueId, CategoryWithinSinglePack> top_level_category_file_locations;
     if (!filesystem::exists(input_path)) {
         cout << "Error: " << input_path << " is not an existing directory or file" << endl;
     }
-
+    map<UniqueId, map<string, vector<string>>> id_conflicts;
     string directory_name = filesystem::path(input_path).filename();
     vector<MarkerPackFile> xml_files = get_files_by_suffix(input_path, ".xml");
     for (const MarkerPackFile& path : xml_files) {
-        set<string> top_level_category_names = parse_xml_file(path, marker_categories, parsed_pois);
+        map<UniqueId, Category*> top_level_categories = parse_xml_file(path, marker_categories, parsed_pois);
         string file_path = join_file_paths(input_path, path.relative_filepath);
-        for (set<string>::iterator it = top_level_category_names.begin(); it != top_level_category_names.end(); it++) {
-            top_level_category_file_locations[*it].push_back(file_path);
+        for (map<UniqueId, Category*>::iterator it = top_level_categories.begin(); it != top_level_categories.end(); it++) {
+            if (top_level_category_file_locations.find(it->first) != top_level_category_file_locations.end()) {
+                if (top_level_category_file_locations[it->first].category != it->second) {
+                    id_conflicts[it->first][top_level_category_file_locations[it->first].category->name] = top_level_category_file_locations[it->first].file_paths;
+                    id_conflicts[it->first][it->second->name].push_back(file_path);
+                    continue;
+                }
+            }
+            else {
+                top_level_category_file_locations[it->first].category = it->second;
+            }
+            top_level_category_file_locations[it->first].file_paths.push_back(file_path);
         }
     }
 
+    if (id_conflicts.size() > 0) {
+        for (map<UniqueId, map<string, vector<string>>>::iterator id_conflicts_it = id_conflicts.begin(); id_conflicts_it != id_conflicts.end(); id_conflicts_it++) {
+            cerr << "Error: Different MarkerCategory nodes were found with same ID (" << id_conflicts_it->first.get_value() << ")" << endl;
+            map<string, vector<string>> category_name_to_file_names;
+            for (map<string, vector<string>>::iterator category_it = id_conflicts_it->second.begin(); category_it != id_conflicts_it->second.end(); category_it++) {
+                for (size_t i = 0; i < category_it->second.size(); i++) {
+                    if (category_it->first == "") {
+                        category_name_to_file_names["not found"].push_back(category_it->second[i]);
```

Comment:
Is this a magic string that we are using here of "not found". Is it possible for a real value to be "not found"?